Kernel/Thread.cpp

```diff
@@ -375,78 +384,53 @@ ShouldUnblockThread Thread::dispatch_signal(u8 signal)
-        kprintf("resume tss pc: %w:%x stack: %w:%x flags: %x cr3: %x\n", m_tss_to_resume_kernel->cs, m_tss_to_resume_kernel->eip, m_tss_to_resume_kernel->ss, m_tss_to_resume_kernel->esp, m_tss_to_resume_kernel->eflags, m_tss_to_resume_kernel->cr3);
-#endif
+    push_value_on_user_stack(regs, ret_eip);
+    push_value_on_user_stack(regs, ret_eflags);
 
-        if (!m_signal_stack_user_region) {
-            m_signal_stack_user_region = m_process.allocate_region(VirtualAddress(), default_userspace_stack_size, String::format("User Signal Stack (Thread %d)", m_tid));
-            ASSERT(m_signal_stack_user_region);
-        }
-        if (!m_kernel_stack_for_signal_handler_region)
-            m_kernel_stack_for_signal_handler_region = MM.allocate_kernel_region(default_kernel_stack_size, String::format("Kernel Signal Stack (Thread %d)", m_tid));
-        m_tss.ss = 0x23;
-        m_tss.esp = m_signal_stack_user_region->vaddr().offset(default_userspace_stack_size).get();
-        m_tss.ss0 = 0x10;
-        m_tss.esp0 = m_kernel_stack_for_signal_handler_region->vaddr().offset(default_kernel_stack_size).get();
-
-        push_value_on_stack(0);
-    } else {
-        push_value_on_stack(ret_eip);
-        push_value_on_stack(ret_eflags);
-
-        // PUSHA
-        u32 old_esp = m_tss.esp;
-        push_value_on_stack(m_tss.eax);
-        push_value_on_stack(m_tss.ecx);
-        push_value_on_stack(m_tss.edx);
-        push_value_on_stack(m_tss.ebx);
-        push_value_on_stack(old_esp);
-        push_value_on_stack(m_tss.ebp);
-        push_value_on_stack(m_tss.esi);
-        push_value_on_stack(m_tss.edi);
-
-        // Align the stack.
-        m_tss.esp -= 12;
-    }
+    // PUSHA
+    push_value_on_user_stack(regs, regs->eax);
+    push_value_on_user_stack(regs, regs->ecx);
+    push_value_on_user_stack(regs, regs->edx);
+    push_value_on_user_stack(regs, regs->ebx);
+    push_value_on_user_stack(regs, old_esp);
+    push_value_on_user_stack(regs, regs->ebp);
+    push_value_on_user_stack(regs, regs->esi);
+    push_value_on_user_stack(regs, regs->edi);
 
     // PUSH old_signal_mask
-    push_value_on_stack(old_signal_mask);
-
-    m_tss.cs = 0x1b;
-    m_tss.ds = 0x23;
-    m_tss.es = 0x23;
-    m_tss.fs = 0x23;
-    m_tss.gs = 0x23;
-    m_tss.eip = handler_vaddr.get();
+    push_value_on_user_stack(regs, old_signal_mask);
 
-    // FIXME: Should we worry about the stack being 16 byte aligned when entering a signal handler?
-    push_value_on_stack(signal);
+    push_value_on_user_stack(regs, signal);
+    push_value_on_user_stack(regs, handler_vaddr.get());
+    push_value_on_user_stack(regs, 0); //push fake return address
 
-    if (interrupting_in_kernel)
-        push_value_on_stack(g_return_to_ring0_from_signal_trampoline.get());
-    else
-        push_value_on_stack(g_return_to_ring3_from_signal_trampoline.get());
+    regs->eip = g_return_to_ring3_from_signal_trampoline.get();
 
-    ASSERT((m_tss.esp % 16) == 0);
```

Comment:
Why are you removing this assertion? Is it okay for the stack to be misaligned?
If so, we should also remove the FIXME comment about stack alignment.